service/src/test/java/bio/terra/pipelines/service/PipelineInputsOutputsServiceTest.java

```diff
@@ -151,4 +166,458 @@ void stringToMapBadString() {
+        allPipelineInputDefinitions.stream()
+            .map(PipelineInputDefinition::getName)
+            .collect(Collectors.toSet())) {
+      assertEquals(expectedPopulatedInputs.get(inputName), populatedInputs.get(inputName));
+    }
+  }
+
+  private static Stream<Arguments> formatPipelineInputsTestValues() {
+    return Stream.of(
+        // arguments: all raw inputs, all pipeline input definitions, inputs with custom values,
+        // keys to prepend with storage url, expected formatted inputs
+        arguments( // no special modifications
+            new HashMap<String, Object>(Map.of("inputName", "user provided value")),
+            new ArrayList<>(
+                List.of(
+                    createTestPipelineInputDef(
+                        PipelineVariableTypesEnum.STRING, true, true, null))),
+            new HashMap<>(), // no inputs with custom values
+            new ArrayList<>(), // no keys to prepend with storage workspace url
+            new HashMap<String, Object>(Map.of("input_name", "user provided value"))),
+        arguments( // overwrite service input value with custom value
+            new HashMap<String, Object>(Map.of("inputName", "will be overwritten")),
+            new ArrayList<>(
+                List.of(
+                    createTestPipelineInputDef(
+                        PipelineVariableTypesEnum.STRING, true, false, "will be overwritten"))),
+            new HashMap<>(Map.of("inputName", "custom value")),
+            new ArrayList<>(), // no keys to prepend with storage workspace url
+            new HashMap<String, Object>(Map.of("input_name", "custom value"))),
+        arguments( // format user file with control workspace url and user input file path
+            new HashMap<String, Object>(Map.of("inputName", "value")),
+            new ArrayList<>(
+                List.of(
+                    createTestPipelineInputDef(PipelineVariableTypesEnum.FILE, true, true, null))),
+            new HashMap<>(), // no inputs with custom values
+            new ArrayList<>(), // no keys to prepend with storage workspace url
+            new HashMap<String, Object>(
+                Map.of(
+                    "input_name",
+                    "gs://control-workspace-bucket/user-input-files/%s/value"
+                        .formatted(TestUtils.TEST_NEW_UUID)))),
+        arguments( // prepend key with storage workspace url
+            new HashMap<String, Object>(Map.of("inputName", "/value")),
+            new ArrayList<>(
+                List.of(
+                    createTestPipelineInputDef(
+                        PipelineVariableTypesEnum.STRING, true, false, null))),
+            new HashMap<>(), // no inputs with custom values
+            new ArrayList<>(List.of("inputName")), // prepend this key with storage workspace url
+            new HashMap<String, Object>(
+                Map.of("input_name", "gs://storage-workspace-bucket/value"))),
+        arguments( // test casting
```

Comment:
it would be nice to have one test where there is more than one item in some of the parameters just to get a more 
"realistic" test

Reply:
We do when we test the parent method, but I can add some in these ones too

Reply:
added a two-input-definition test case here and in the previous set